Source/WebCore/loader/NavigationScheduler.cpp

```diff
@@ -315,6 +316,60 @@ class ScheduledHistoryNavigation : public ScheduledNavigation {
     Ref<HistoryItem> m_historyItem;
 };
 
+class ScheduledHistoryNavigationByKey : public ScheduledNavigation {
+public:
+    explicit ScheduledHistoryNavigationByKey(const String& key, CompletionHandler<void(ScheduleHistoryNavigationResult)>&& completionHandler)
+        : ScheduledNavigation(0, LockHistory::No, LockBackForwardList::No, false, true)
+        , m_key(key)
+        , m_completionHandler(WTFMove(completionHandler))
+    {
+    }
+
+    ~ScheduledHistoryNavigationByKey()
+    {
+        if (m_completionHandler)
+            m_completionHandler(ScheduleHistoryNavigationResult::Aborted);
+    }
+
+    void fire(Frame& frame) override
+    {
+        RefPtr localFrame = dynamicDowncast<LocalFrame>(frame);
+        // If the destination HistoryItem is no longer in the back/forward list, then we don't proceed.
+        RefPtr page { frame.page() };
+        if (!page || !localFrame) {
+            m_completionHandler(ScheduleHistoryNavigationResult::Aborted);
+            return;
+        }
+
+        UserGestureIndicator gestureIndicator(userGestureToForward());
+
+        auto entry = localFrame->window()->navigation().findEntryByKey(m_key);
+        if (!entry) {
+            m_completionHandler(ScheduleHistoryNavigationResult::Aborted);
+            return;
+        }
+        Ref historyItem = entry.value()->associatedHistoryItem();
+
+        if (page->backForward().currentItem() == historyItem.ptr()) {
+            // Special case for go(0) from a frame -> reload only the frame
+            // To follow Firefox and IE's behavior, history reload can only navigate the self frame.
+            if (RefPtr localFrame = dynamicDowncast<LocalFrame>(frame))
+                localFrame->checkedLoader()->changeLocation(localFrame->document()->url(), selfTargetFrameName(), 0, ReferrerPolicy::EmptyString, shouldOpenExternalURLs(), std::nullopt, nullAtom(), std::nullopt, NavigationHistoryBehavior::Reload);
+            return;
+        }
+
+        // go(i!=0) from a frame navigates into the history of the frame only,
+        // in both IE and NS (but not in Mozilla). We can't easily do that.
```

Comment:
Netscape didn't implement the new navigation API.  I think we should remove this comment.  Same for the comment on 354.  If it is intended to share code, it should be shared rather than copied.

Reply:
I removed the comments.